cli/src/pcluster/models/cluster_config.py

```diff
@@ -1350,29 +1352,105 @@ def _validate(self):
                 )
 
 
-class ClusterBucket:
+# ------------ Attributes class used in cluster resources ----------- #
+
+
+class ClusterBucket(S3Bucket):
     """Represent the cluster s3 bucket configuration."""
 
     def __init__(
         self,
-        name: str,
-        artifact_directory: str,
-        remove_on_deletion: bool,
+        cluster_name: str,
+        stack_name: str,
+        name: str = None,
+        artifact_remove_on_deletion: bool = True,
     ):
-        super().__init__()
-        self.name = name
-        self.artifact_directory = artifact_directory
-        self.remove_on_deletion = remove_on_deletion
+        super().__init__(artifact_remove_on_deletion, name)
+        self.cluster_name = cluster_name
+        self.stack_name = stack_name
+        self.root_cluster_directory = "clusters"
+        self.source_config_name = "cluster-config-original.yaml"
+        self.config_name = "cluster-config.yaml"
+        self.template_name = "aws-parallelcluster.cfn.yaml"
+        self.artifacts_name = "artifacts.zip"
+        self.instance_types_data_name = "instance-types-data.json"
+        self.cluster_directory = None
+        self.full_cluster_dir = None
+        self._get_cluster_dir()
+
+    def _get_object_key(self, object_type, object_name):
+        """Get object key of an artifact."""
+        return os_lib.path.join(self._get_cluster_dir(), object_type, object_name)
+
+    def _get_cluster_dir(self):
+        """Get cluster directory in S3 bucket by stack output."""
+        if self.full_cluster_dir is None:
+            try:
+                stack_data = AWSApi.instance().cfn.describe_stack(self.stack_name)
+                self.full_cluster_dir = next(
+                    (
+                        out["OutputValue"]
+                        for out in stack_data.get("Outputs", [])
+                        if out["OutputKey"] == "ArtifactS3RootDirectory"
+                    ),
+                    None,
+                )
+                self.cluster_directory = self.full_cluster_dir.split("/")[-1]
+                return self.full_cluster_dir
+            except (AWSClientError, KeyError) as e:
+                if f"Stack with id {self.stack_name} does not exist" in str(e):
+                    self._generate_cluster_dir()
+                else:
+                    LOGGER.error("No cluster dir found in stack output.")
+                    raise
```

Comment:
This causes a KeyError to be propagated that is very hard to debug or handle in the caller. Shall we remove the KeyError from the except and handle this case with some validation + raising the proper exception if needed?